Crash when recycling Bitmaps too quickly

<!--
Please fill in the below fields with some data to help us best diagnose the issue.
The more specific you are, the better! You can help a lot by not making us ask these questions.
Feel free to remove any irrelevant parts that you know are not related to the issue.
Any HTML comment like this will be stripped when rendering markdown, no need to delete them.
-->


<!-- What version of Glide you're running, for example: 3.7.1 | 3.8.0-SNAPSHOT | 4.0.0-SNAPSHOT
It's essentially the version number from your build.gradle: `dependencies { compile '...:x.y.z' }` -->
**Glide Version**: 4.2.0

<!-- Do you use any integration library, like OkHttp3 or Volley? For example:
Fails to display with stock networking, but works with okhttp3-1.4.0 -->
**Integration libraries**: okhttp3-3.8.1

<!-- What devices you managed to get the issue to come up on? For example:
fails on Galaxy S4/GT-I9500 4.4.2, works fine on Nexus 6P 5.1 and Genymotion Nexus 5 5.0.1 -->
**Device/Android Version**: Emulator API Level 23, Kindle Fire Android 5.1

<!-- Share the details of your issue in prose, detailing actual and expected behavior. It also helps if you give some info **why** you are trying to do something as opposed to **what** is not working. -->
**Issue details / Repro steps / Use case background**: 
I have a RecyclerView full of LinearLayouts that contain largish ImageViews. They load asynchronously from the network. If I scroll fast enough after a while the app will crash.

<!-- How do you use Glide?
Make sure you include everything as is in your app's code:
Changing a single method parameter can yield totally different results.
Please clarify any magic variables that appear in the code, for example: "// `this` is a Fragment"
-->

**ModelLoader**:
```java
public class BGChartModelLoader extends BaseGlideUrlLoader<BGChart> {
    private final float mScale;

    protected BGChartModelLoader(ModelLoader<GlideUrl, InputStream> concreteLoader) {
        super(concreteLoader);

        final DisplayMetrics metrics = Resources.getSystem().getDisplayMetrics();
        mScale = metrics.density;
    }

    /**
     * The default factory for {@link }de.boersego.gmt.android.glide.BGChartModelLoader
     */
    public static class Factory implements ModelLoaderFactory<BGChart, InputStream> {

        @Override
        public ModelLoader<BGChart, InputStream> build(MultiModelLoaderFactory multiFactory) {
            return new BGChartModelLoader(multiFactory.build(GlideUrl.class, InputStream.class));
        }

        @Override
        public void teardown() {
        }
    }

    @Override
    protected String getUrl(final BGChart chart, final int width, final int height, final Options options) {
        final String locale = options.get(MyrGlideModule.OPTION_LOCALE);
        Float scaleOption = options.get(MyrGlideModule.OPTION_SCALE);
        final float scale = scaleOption != null ? scaleOption : mScale;

        return "http://charting.godmode-trader.de/t/app/config?id=" + chart.id + "&width=" + (width / mScale) + "&height=" + (height / mScale) + "&ratio=" + scale + "&_locale=" + locale;
    }

    @Override
    public boolean handles(BGChart o) {
        // We handle any Chart for now
        return true;
    }
}
```

**Glide load line in RelativeLayout subclass (list cell)**:
```java
            GlideApp.with(this).load((BGChart)mData)
                    .apply(RequestOptions.option(MyrGlideModule.OPTION_LOCALE, PanelDataSource.getInstance(getContext()).getLanguageCode()))
                    .error(R.drawable.ic_error_yellow_800_24dp)
                    .placeholder(new ActivityIndicatorDrawable(getContext()))
                    .into(mChartThumb);
```

**GlideModule**:
```java
@GlideModule
public class MyrGlideModule extends AppGlideModule {

    public static final Option<String> OPTION_LOCALE = Option.disk("locale", "en", new Option.CacheKeyUpdater<String>() {
        @Override
        public void update(byte[] keyBytes, String value, MessageDigest messageDigest) {
            messageDigest.update(keyBytes);
            messageDigest.update(value.getBytes());
        }
    });

    public static final Option<Float> OPTION_SCALE = Option.disk("scale", new Option.CacheKeyUpdater<Float>() {
        @Override
        public void update(byte[] keyBytes, Float value, MessageDigest messageDigest) {
            messageDigest.update(keyBytes);
            messageDigest.update(String.valueOf(value).getBytes());
        }
    });
    private static CircleBorderTransformation sCircleBorderTransform;

    @Override
    public void applyOptions(Context context, GlideBuilder builder) {
        // Standard Glide options for the whole app. We want a smaller disk cache since we won't be loading big images
        // outside of WebView (which has it's own caching), and also set log level to high for non-debug builds
        builder.setLogLevel(BuildConfig.DEBUG ? Log.VERBOSE : Log.WARN);
        int diskCacheSizeBytes = 1024 * 1024 * 100; // 100 MB
        builder.setDiskCache(new InternalCacheDiskCacheFactory(context, diskCacheSizeBytes));
        // Initialize circle border transform paint from resources
    }

    @Override
    public void registerComponents(Context context, Glide glide, Registry registry) {
        registry.append(BGUser.class, InputStream.class, new AvatarModelLoader.Factory());
        registry.append(BGChart.class, InputStream.class, new BGChartModelLoader.Factory());
        registry.append(BGCountry.class, InputStream.class, new CountryModelLoader.Factory());
    }


}
```

<!-- How does your app look like?
We're most interested in the layout attributes and the hierarchy around the ImageView -->
**Layout XML**:
```xml
<?xml version="1.0" encoding="utf-8"?>
<android.support.v7.widget.CardView xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:card_view="http://schemas.android.com/apk/res-auto"
    xmlns:tools="http://schemas.android.com/tools"
    android:id="@+id/chart_card"
    android:layout_width="match_parent"
    android:layout_height="match_parent"
    android:duplicateParentState="true"
    card_view:cardCornerRadius="3dp"
    card_view:cardElevation="2dp"
    card_view:cardUseCompatPadding="true">

    <RelativeLayout
        android:layout_width="match_parent"
        android:layout_height="wrap_content"
        android:background="@drawable/bg_selector_action_mode">

        <TextView
            android:id="@+id/chart_name"
            style="@style/font_bold"
            android:layout_width="wrap_content"
            android:layout_height="wrap_content"
            android:layout_alignParentStart="true"
            android:layout_alignParentTop="true"
            android:layout_marginStart="11dp"
            android:layout_marginTop="11dp"
            android:layout_toStartOf="@+id/chart_date"
            android:ellipsize="end"
            android:maxLines="1"
            android:textColor="@color/veryDarkText"
            android:textIsSelectable="false"
            android:textSize="@dimen/TextStyleMedium"
            tools:text="Chart Analyse 2017" />

        <TextView
            android:id="@+id/chart_date"
            android:layout_width="wrap_content"
            android:layout_height="wrap_content"
            android:layout_alignParentEnd="true"
            android:layout_alignParentTop="true"
            android:layout_marginEnd="11dp"
            android:layout_marginTop="11dp"
            android:maxLines="1"
            android:textColor="@color/bgo_grey"
            android:textIsSelectable="false"
            android:textSize="@dimen/TextStyleSmall"
            tools:text="19.10.2015" />

        <TextView
            android:id="@+id/chart_tags"
            android:layout_width="match_parent"
            android:layout_height="wrap_content"
            android:layout_below="@+id/chart_name"
            android:layout_marginEnd="11dp"
            android:layout_marginStart="11dp"
            android:textColor="@color/bgo_grey"
            android:textIsSelectable="false"
            android:textSize="@dimen/TextStyleSmall"
            tools:text="Tags: Tag 1, Tag 2, Tag 3" />

        <ImageView
            android:id="@+id/chart_thumbnail"
            android:layout_width="match_parent"
            android:layout_height="195dp"
            android:layout_alignParentStart="true"
            android:layout_below="@+id/chart_tags"
            android:layout_marginTop="13dp"
            android:scaleType="centerCrop" />

    </RelativeLayout>

</android.support.v7.widget.CardView>
```

<!--
What is the error message that you got in the log?
You can find some help on diagnosing issues here: https://github.com/bumptech/glide/wiki/Debugging-and-Error-Handling
-->
**Stack trace / LogCat**:
```ruby
                         E  java.lang.IllegalStateException: Cannot recycle a resource that has already been recycled
                         E      at com.bumptech.glide.load.engine.EngineResource.recycle(EngineResource.java:60)
                         E      at com.bumptech.glide.load.engine.ResourceRecycler.recycle(ResourceRecycler.java:28)
                         E      at com.bumptech.glide.load.engine.Engine.onResourceRemoved(Engine.java:311)
                         E      at com.bumptech.glide.load.engine.cache.LruResourceCache.onItemEvicted(LruResourceCache.java:31)
                         E      at com.bumptech.glide.load.engine.cache.LruResourceCache.onItemEvicted(LruResourceCache.java:11)
                         E      at com.bumptech.glide.util.LruCache.trimToSize(LruCache.java:175)
                         E      at com.bumptech.glide.util.LruCache.evict(LruCache.java:180)
                         E      at com.bumptech.glide.util.LruCache.put(LruCache.java:135)
                         E      at com.bumptech.glide.load.engine.cache.LruResourceCache.put(LruResourceCache.java:11)
                         E      at com.bumptech.glide.load.engine.Engine.onResourceReleased(Engine.java:319)
                         E      at com.bumptech.glide.load.engine.EngineResource.release(EngineResource.java:101)
                         E      at com.bumptech.glide.load.engine.Engine.release(Engine.java:277)
                         E      at com.bumptech.glide.request.SingleRequest.releaseResource(SingleRequest.java:324)
                         E      at com.bumptech.glide.request.SingleRequest.clear(SingleRequest.java:303)
                         E      at com.bumptech.glide.manager.RequestTracker.clearRemoveAndRecycle(RequestTracker.java:62)
                         E      at com.bumptech.glide.RequestManager.untrack(RequestManager.java:443)
                         E      at com.bumptech.glide.RequestManager.untrackOrDelegate(RequestManager.java:430)
                         E      at com.bumptech.glide.RequestManager.clear(RequestManager.java:418)
                         E      at com.bumptech.glide.RequestBuilder.into(RequestBuilder.java:380)
                         E      at com.bumptech.glide.RequestBuilder.into(RequestBuilder.java:432)
                         E      at de.boersego.gmt.android.ui.cells.ChartCell.run(ChartCell.java:101)
                         E      at de.boersego.gmt.android.ui.cells.ChartCell.setData(ChartCell.java:86)
                         E      at de.boersego.gmt.android.base.BGRecycleListAdapter.onBindViewHolder(BGRecycleListAdapter.java:451)
                         E      at de.boersego.gmt.android.base.BGRecycleListAdapter.onBindViewHolder(BGRecycleListAdapter.java:44)
                         E      at android.support.v7.widget.RecyclerView$Adapter.onBindViewHolder(RecyclerView.java:6482)
                         E      at android.support.v7.widget.RecyclerView$Adapter.bindViewHolder(RecyclerView.java:6515)
                         E      at android.support.v7.widget.RecyclerView$Recycler.tryBindViewHolderByDeadline(RecyclerView.java:5458)
                         E      at android.support.v7.widget.RecyclerView$Recycler.tryGetViewHolderForPositionByDeadline(RecyclerView.java:5724)
                         E      at android.support.v7.widget.GapWorker.prefetchPositionWithDeadline(GapWorker.java:285)
                         E      at android.support.v7.widget.GapWorker.flushTaskWithDeadline(GapWorker.java:342)
                         E      at android.support.v7.widget.GapWorker.flushTasksWithDeadline(GapWorker.java:358)
                         E      at android.support.v7.widget.GapWorker.prefetch(GapWorker.java:365)
                         E      at android.support.v7.widget.GapWorker.run(GapWorker.java:396)
                         E      at android.os.Handler.handleCallback(Handler.java:739)
                         E      at android.os.Handler.dispatchMessage(Handler.java:95)
                         E      at android.os.Looper.loop(Looper.java:135)
                         E      at android.app.ActivityThread.main(ActivityThread.java:5484)
                         E      at java.lang.reflect.Method.invoke(Native Method)
                         E      at java.lang.reflect.Method.invoke(Method.java:372)
                         E      at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:984)
                         E      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:779)
                         I  To Report FATAL to activityManagerService
        ActivityManager  I  handleApplicationCrash
                         W    Force finishing activity 1 de.boersego.gmt.android/.MainActivityOverride
```

<!-- Bonus points if you attach a relevant screenshot, screen recording or a small demo project -->

--- a/library/src/main/java/com/bumptech/glide/util/LruCache.java
+++ b/library/src/main/java/com/bumptech/glide/util/LruCache.java
@@ -169,7 +169,7 @@ public class LruCache<T, Y> {
   protected synchronized void trimToSize(int size) {
     Map.Entry<T, Y> last;
     final Iterator<Map.Entry<T, Y>> cacheIterator = cache.entrySet().iterator();
-    while (currentSize > size) {
+    while (currentSize > size && cacheIterator.hasNext()) {
       last = cacheIterator.next();
       final Y toRemove = last.getValue();
       currentSize -= getSize(toRemove);

--- a/library/src/main/java/com/bumptech/glide/util/LruCache.java
+++ b/library/src/main/java/com/bumptech/glide/util/LruCache.java
@@ -1,6 +1,8 @@
 package com.bumptech.glide.util;
 
 import android.support.annotation.Nullable;
+
+import java.util.Iterator;
 import java.util.LinkedHashMap;
 import java.util.Map;
 
@@ -166,12 +168,13 @@ public class LruCache<T, Y> {
    */
   protected synchronized void trimToSize(int size) {
     Map.Entry<T, Y> last;
+    final Iterator<Map.Entry<T, Y>> cacheIterator = cache.entrySet().iterator();
     while (currentSize > size) {
-      last = cache.entrySet().iterator().next();
+      last = cacheIterator.next();
       final Y toRemove = last.getValue();
       currentSize -= getSize(toRemove);
       final T key = last.getKey();
-      cache.remove(key);
+      cacheIterator.remove();
       onItemEvicted(key, toRemove);
     }
   }

--- a/library/src/main/java/com/bumptech/glide/util/LruCache.java
+++ b/library/src/main/java/com/bumptech/glide/util/LruCache.java
@@ -168,8 +168,9 @@ public class LruCache<T, Y> {
    */
   protected synchronized void trimToSize(int size) {
     Map.Entry<T, Y> last;
-    final Iterator<Map.Entry<T, Y>> cacheIterator = cache.entrySet().iterator();
-    while (currentSize > size && cacheIterator.hasNext()) {
+    Iterator<Map.Entry<T, Y>> cacheIterator;
+    while (currentSize > size) {
+      cacheIterator  = cache.entrySet().iterator();
       last = cacheIterator.next();
       final Y toRemove = last.getValue();
       currentSize -= getSize(toRemove);

